pallets/parachain-staking/src/lib.rs

```diff
@@ -438,9 +445,51 @@ pub mod pallet {
 				weight = weight.saturating_add(T::OnNewRound::on_new_round(round.current));
 				// pay all stakers for T::RewardPaymentDelay rounds ago
 				weight = weight.saturating_add(Self::prepare_staking_payouts(round.current));
+
+				// select total collators
+				let collators = <SelectedCandidates<T>>::get();
+
+				// if collators length is below or eq to 66% of max_collators,
+				// we don't mark any other collator as offline
+				let max_collators = <TotalSelected<T>>::get();
+
+				// track collators length
+				let mut len_counter = collators.len();
+
+				// iter collators to check which of them must be marked as offline
+				for collator in collators {
+					if let Some(max_offline_counter) = <CandidateLastActive<T>>::get(&collator) {
+						if len_counter * 3 > (max_collators * 2) as usize
+							&& max_offline_counter >= T::MaxOfflineRounds::get()
+						{
+							// if the collator has not produced any block within
+							// MaxOfflineRounds e.g(3 rounds for Moonriver)
+							// it is marked as offline
+							let mark_offline_weight = T::MarkOfflineCallback::mark_offline(
+								collator.clone(),
+								round.current.saturating_sub(1),
+							);
+							weight = weight.saturating_add(mark_offline_weight);
+							len_counter = len_counter.saturating_sub(1);
```

Comment:
With the new generic callback, it's wrong top assume that the collator was marked offline, it may not be the case, the callback should return a more complex type (create a new struct for that) to express if a collator should be marked offline or not

Reply:
Maybe I used the wrong name for the callback and its function. I'm not assuming that a collator is going to be marked as offline. Inside the runtime and using the callback, I check that the provided collator is not an orbiter. If it's not an orbiter, I mark it as offline, but any other behavior could be implemented. The code implemented for the callback within the runtime is: 

```rust
pub struct CustomMarkOfflineCallback;
impl pallet_parachain_staking::MarkOfflineCallback<Runtime> for CustomMarkOfflineCallback {
	fn mark_offline(collator_id: AccountId, round: pallet_parachain_staking::RoundIndex) -> Weight {
		let extra_weight = if !MoonbeamOrbiters::is_orbiter(round, collator_id.clone()) {
			ParachainStaking::do_go_offline(collator_id).unwrap_or_default()
		} else {
			Weight::zero()
		};

		<Runtime as frame_system::Config>::DbWeight::get()
			.reads(1)
			.saturating_add(extra_weight)
	}
}
``` 
So, we could implement any other behavior instead of calling `do_go_offline()` to deal with the inactive collator. I will rename the callback to avoid confusions by the way.

Reply:
> I'm not assuming that a collator is going to be marked as offline

Decrementing the `len_counter` mean that the pallet assume that the callback should mark the collactor as offline, that's why I pointed out this particular line

>  The code implemented for the callback within the runtime is:

The code you provide as an example is not valid, in the particular case of moonbeam, our implementation of the callback should call a new function in the orbiter pallet (still to be coded), this function should check if the collator is a parent collator that has at least 1 active orbiter, if so deactivate this particular orbiter and return a value saying to the staking pallet that the collator should not be marked offline, otherwise do nothing and tell the pallet that the collator should be marked offline

It's even more complex than that, because the behavior I describe is only valid if `MaxOfflineRounds == RotatePeriod`

Reply:
Okay I understand now. I'll be checking this